Consensus logging

VoidingWarranties opened the pull request:
Adds basic logging to consensus. None of the logging statements are wrapped in a `if build.DEBUG` as they are all logging errors that should not occur frequently.

VoidingWarranties commented on modules/consensus/consensusset.go:
```diff
@@ -174,9 +175,13 @@ func (cs *ConsensusSet) ChildTarget(id types.BlockID) (target types.Target, exis
 func (cs *ConsensusSet) Close() error {
+	// TODO: is this lock necessary?
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
```
Is it necessary or good practice to hold a lock during Close?

DavidVorick replied:
if you are changing things, you should have a lock.

DavidVorick replied:
furthermore, `Close` is exported, so people will be assuming that there's a lock. That doesn't necessarily mean that it needs a lock, but if it's a larger function it should have one so that devs don't make the wrong assumptions.

DavidVorick commented:
We definitely need some call to logging that will only log the message in DEBUG mode.

VoidingWarranties commented:
Agreed, more logging in DEBUG mode is good. I'm not very familiar with consensus code outside of synchronizing, so I don't know where else logging might be useful.

DavidVorick commented on modules/consensus/accept.go:
```diff
@@ -237,8 +237,10 @@ func (cs *ConsensusSet) managedAcceptBlock(b types.Block) error {
-					cs.AcceptBlock(b) // NOTE: Error is not handled.
+					err := cs.AcceptBlock(b)
+					if err != nil {
+						cs.log.Println("WARN: failed to accept a future block:", err)
```
this should be debug logged only.

DavidVorick commented on modules/consensus/consensusset.go:
```diff
@@ -176,7 +177,10 @@ func (cs *ConsensusSet) ChildTarget(id types.BlockID) (target types.Target, exis
+	return build.JoinErrors([]error{
+		build.AddContext("log.Close failed:", cs.log.Close()),
+		build.AddContext("db.Close failed:", cs.db.Close()),
+	}, "; ")
```
This is interesting, I'm not sure I like the construction. I like both the context and the fact that we're joining the errors instead of returning early, but not necessarily the actual way the code is constructed. You are returning a function that's taking an array of function calls that have function calls as arguments.

Doesn't seem like the right way to do things.

Also, if you put the last argument of `JoinErrors` as the first argument, you could make the array variadic, which I think is more appropriate for `JoinErrors` anyway.

lukechampine replied:
Variadic appeals to me too, although `build.JoinErrors` currently has nice symmetry with `strings.Join`.

Ooh, something else occurred to me: if `build.JoinErrors` only sees one non-nil error, it should return that error directly instead of creating a new error. This preserves the ability to check for things like `io.EOF`.

VoidingWarranties replied:
>Ooh, something else occurred to me: if `build.JoinErrors` only sees one non-nil error, it should return that error directly instead of creating a new error. This preserves the ability to check for things like `io.EOF`.

That seems nice in theory, but I don't see the practical advantage. It encourages clients of `JoinErrors` or functions that use `JoinErrors` to rely on the behavior and trust that there is only one error being returned.

lukechampine replied:
hmm. Fair point. It seems like most of these joined errors go straight to logs or to the user anyway.

VoidingWarranties replied:
A disadvantage of the `build.JoinErrors` pattern here is that you can only call functions that only return an error, and you cannot intersperse other function calls (such as locking / unlocking).

Then again, maybe it is good style to enforce that the components of the module be able to close easily like this, as it makes it easier to reason about.

DavidVorick commented on modules/consensus/synchronize.go:
```diff
@@ -327,9 +327,12 @@ func (cs *ConsensusSet) rpcRelayBlock(conn modules.PeerConn) error {
+		go func() {
+			err := cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
+			if err != nil {
+				cs.log.Println("WARN: failed to get parents of orphan block:", err)
```
definitely needs to be logged only if DEBUG

DavidVorick commented on modules/consensus/synchronize.go:
```diff
@@ -355,18 +358,24 @@ func (cs *ConsensusSet) rpcRelayHeader(conn modules.PeerConn) error {
+		go func() {
+			err := cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
+			if err != nil {
+				cs.log.Println("WARN: failed to get parents of orphan header:", err)
```
same here

DavidVorick commented on modules/consensus/synchronize.go:
```diff
@@ -355,18 +358,24 @@ func (cs *ConsensusSet) rpcRelayHeader(conn modules.PeerConn) error {
+	go func() {
+		err := cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlk", cs.threadedReceiveBlock(h.ID()))
+		if err != nil {
+			cs.log.Println("WARN: failed to get header's corresponding block:", err)
```
aaaand same here

DavidVorick commented on modules/consensus/synchronize.go:
```diff
@@ -433,8 +442,9 @@ func (cs *ConsensusSet) threadedInitialBlockchainDownload() {
+	numOutboundSynced := 0
 	for {
-		numOutboundSynced := 0
+		numOutboundSynced = 0
```
can you explain why you moved `numOutboudSynced` outside the loop? I'm guessing it was a bug?

VoidingWarranties replied:
Not a bug, it's needed for the last line outside the for loop
```go
cs.log.Printf("INFO: IBD done, synced with %v peers", numOutboundSynced)
```

DavidVorick commented:
As you can see from my comments, pretty picky about what types of logging statements should be making production. Basically, only statements which indicate that the developer has done something wrong (like peers failing to disconnect) or that the user machine is having errors. Things that are triggered by bad peers, etc. shouldn't be logged unless in DEBUG mode.

One-time statements such as 'IBD complete with x peers' also make sense.

VoidingWarranties commented:
I added `Logger.Debug`, `Logger.Debugf`, and `Logger.Debugln`. There's not testing for them as `build.DEBUG` is a constant that's always true during testing.

I changed the noted log statements to use `Debugln`.

VoidingWarranties commented:
Build is failing because of https://github.com/travis-ci/travis-ci/issues/5806

DavidVorick commented:
Don't we already only have a `Println` for the logger? I don't think adding all 3 was necessary, though strictly speaking I don't think it's harmful
